File: math/60.cpp
#include<vector>
#include<iostream>
#include<string>

using namespace std; 

class Solution {
public:
    int s1 = 1;
    string getPermutation(int n, int k) {
        string result="";
        vector<string> v(n);
        for(int i=0; i<n; i++) v[i]=(char)(i+1 + '0');
        if(n == 1) return "1";
        int sn=1, tmp=n;
        while(tmp > 0) sn *= tmp--;
        rec(result, v, n, k, sn);
        return result;

    }
    void rec(string& result, vector<string> v, int n, int k, int sn){
        if(n == 2){
            result = result + v[k/2];
            v.erase(v.begin() + k/2);
            result = result + v[0];
            return;
        }else{
            int sn1 = sn / n,  posi = k/sn1, res = k%sn1;
            if(res == 0) posi--, res=sn1; 
            result = result + v[posi];
            v.erase(v.begin() + posi);
            rec(result, v, n-1, res, sn1);
        }
    }
};


int main(){ 
    int n = 3;
    int k = 6;
    Solution s;
    string result = s.getPermutation(n, k);
    cout<<result<<endl;
    return -1;
}

File: recursive/60.py
#-*-coding:utf-8-*-
'''
@Author: llei
@Date: 2022-05-30 12:40:03
'''
from curses.ascii import SO
import sys
import os
from typing import List, Tuple, Dict

sys.path.append("..")

class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        
        result = []
        stop = False
        nums = list(range(1, n+1))

        def dfs(tmp, nums):
            nonlocal stop
            if len(nums) == 0:
                result.append(tmp)
                if len(result) == k:
                    stop = True
                return
            else:
                if stop:
                    return
                for i, n in enumerate(nums):
                    dfs(tmp + [n], nums[:i] + nums[i+1:])
                    if stop:
                        return

        dfs([], nums)

        return "".join([str(t) for t in result[-1]])
        
if __name__ == "__main__":
    module = Solution()
    module.getPermutation(4, 9)